src/awkward/operations/structure.py

```diff
@@ -2157,6 +2157,202 @@ def getfunction(layout, depth, posaxis):
+
+            # Compact indices
+            for i in range(len(layout.contents)):
+                is_i = tags == i
+
+                new_contents[i] = apply(layout.project(i), depth, posaxis)
+                new_index[is_i] = nplike.arange(nplike.sum(is_i))
+
+            return ak.layout.UnionArray8_64(
+                ak.layout.Index8(tags),
+                ak.layout.Index64(new_index),
+                new_contents,
+                layout.identities,
+                layout.parameters,
+            )
+
+        # RecordArray contents can be truncated
+        if isinstance(layout, ak.layout.RecordArray):
+            return ak.layout.RecordArray(
+                [
+                    apply(truncate(c, len(layout)), depth, posaxis)
+                    for c in layout.contents
+                ],
+                layout.recordlookup,
+                len(layout),
+                layout.identities,
+                layout.parameters,
+            )
+
+        # RegularArrays can change length
+        if isinstance(layout, ak.layout.RegularArray):
+            if not len(layout):
+                return layout
+
+            content = layout.content
+
+            # Truncate content if it is larger than a perfect
+            # multiple of the RegularArray size
+            n, r = divmod(len(content), layout.size)
+            if r:
+                content = truncate(content, n * layout.size)
+
+            return ak.layout.RegularArray(
+                apply(content, depth + 1, posaxis), layout.size
+            )
+
+        # BitMaskedArrays can change length
+        if isinstance(layout, ak.layout.BitMaskedArray):
+            layout = layout.simplify()
+
+            if not isinstance(ak.type(layout.content), ak.types.PrimitiveType):
+                return apply(layout.toIndexedOptionArray64(), depth, posaxis)
+
+            return ak.layout.BitMaskedArray(
+                layout.mask,
+                apply(truncate(layout.content, len(layout)), depth, posaxis),
+                layout.valid_when,
+                len(layout),
+                layout.lsb_order,
+                layout.identities,
+                layout.parameters,
+            )
+
+        # ByteMaskedArrays can change length
+        if isinstance(layout, ak.layout.ByteMaskedArray):
+            layout = layout.simplify()
+
+            if not isinstance(ak.type(layout.content), ak.types.PrimitiveType):
+                return apply(layout.toIndexedOptionArray64(), depth, posaxis)
+
+            return ak.layout.ByteMaskedArray(
+                layout.mask,
+                apply(truncate(layout.content, len(layout)), depth, posaxis),
+                layout.valid_when,
+                layout.identities,
+                layout.parameters,
+            )
+
+        if isinstance(layout, ak.layout.VirtualArray):
+            return apply(layout.array, depth, posaxis)
+
+        # Finally, fall through to failure
+        raise NotImplementedError
```

Comment:
Since the above should be an `elif` chain, this would be the `else` clause, and it should actually be an `AssertionError` (whose argument is the actual `layout` or `type(layout)`, for debugging) as in other such chains. Missing a case is considered a programming error: it's not an open-ended set—these are all the node types.